To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [blog post](https://unsloth.ai/blog/r1-reasoning) for guidance on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
# Skip restarting message in Colab
import sys; modules = list(sys.modules.keys())
for x in modules: sys.modules.pop(x) if "PIL" in x or "google" in x else None

!pip install unsloth vllm
!pip install --upgrade pillow
# If you are running this notebook on local, you need to install `diffusers` too
# !pip install diffusers
# Temporarily install a specific TRL nightly version
!pip install git+https://github.com/huggingface/trl.git@e95f9fb74a3c3647b86f251b7e230ec51c64b72b

### Unsloth

Use `PatchFastRL` before all functions to patch GRPO and other RL algorithms!

In [2]:
from unsloth import FastLanguageModel, PatchFastRL
PatchFastRL("GRPO", FastLanguageModel)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
INFO 02-12 01:25:48 __init__.py:190] Automatically detected platform cuda.


Load up `Llama 3.1 8B Instruct`, and set parameters

In [3]:
from unsloth import is_bfloat16_supported
import torch
max_seq_length = 20000 # Can increase for longer reasoning traces
lora_rank = 64 # Larger rank = smarter, but slower

model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "meta-llama/meta-Llama-3.1-8B-Instruct",
    max_seq_length = max_seq_length,
    load_in_4bit = True, # False for LoRA 16bit
    fast_inference = True, # Enable vLLM fast inference
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.4, # Reduce if out of memory
)

model = FastLanguageModel.get_peft_model(
    model,
    r = lora_rank, # Choose any number > 0 ! Suggested 8, 16, 32, 64, 128
    target_modules = [
        "gate_proj", "up_proj", "down_proj",
    ], # Remove QKVO if out of memory
    lora_alpha = lora_rank,
    use_gradient_checkpointing = "unsloth", # Enable long context finetuning
    random_state = 3407,
)

==((====))==  Unsloth 2025.2.5: Fast Llama patching. Transformers: 4.48.2.
   \\   /|    GPU: NVIDIA A100-SXM4-40GB. Max memory: 39.557 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.5.1+cu124. CUDA: 8.0. CUDA Toolkit: 12.4. Triton: 3.1.0
\        /    Bfloat16 = TRUE. FA [Xformers = 0.0.28.post3. FA2 = False]
 "-____-"     Free Apache license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: vLLM loading unsloth/meta-llama-3.1-8b-instruct-bnb-4bit with actual GPU utilization = 39.58%
Unsloth: Your GPU has CUDA compute capability 8.0 with VRAM = 39.56 GB.
Unsloth: Using conservativeness = 1.0. Chunked prefill tokens = 20000. Num Sequences = 224.
Unsloth: vLLM's KV Cache can use up to 9.33 GB. Also swap space = 6 GB.
INFO 02-12 01:26:07 config.py:542] This model supports multiple tasks: {'reward', 'score', 'generate', 'classify', 'embed'}. Defaulting to 'generate'.
Unsloth: vLLM Bitsandbytes config using kwar

Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


INFO 02-12 01:26:14 model_runner.py:1115] Loading model weights took 5.3541 GB
INFO 02-12 01:26:14 punica_selector.py:18] Using PunicaWrapperGPU.
INFO 02-12 01:26:18 worker.py:267] Memory profiling takes 4.08 seconds
INFO 02-12 01:26:18 worker.py:267] the current vLLM instance can use total_gpu_memory (39.56GiB) x gpu_memory_utilization (0.40) = 15.66GiB
INFO 02-12 01:26:18 worker.py:267] model weights take 5.35GiB; non_torch_memory takes 0.09GiB; PyTorch activation peak memory takes 2.27GiB; the rest of the memory reserved for KV Cache is 7.94GiB.
INFO 02-12 01:26:19 executor_base.py:110] # CUDA blocks: 4064, # CPU blocks: 3072
INFO 02-12 01:26:19 executor_base.py:115] Maximum concurrency for 20000 tokens per request: 3.25x
INFO 02-12 01:26:23 model_runner.py:1434] Capturing cudagraphs for decoding. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI. If out-of-memory error oc

Capturing CUDA graph shapes: 100%|██████████| 31/31 [00:44<00:00,  1.43s/it]

INFO 02-12 01:27:08 model_runner.py:1562] Graph capturing finished in 44 secs, took 0.77 GiB
INFO 02-12 01:27:08 llm_engine.py:431] init engine (profile, create kv cache, warmup model) took 53.74 seconds



Not an error, but Unsloth cannot patch Attention layers with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Not an error, but Unsloth cannot patch O projection layer with our manual autograd engine since either LoRA adapters
are not enabled or a bias term (like in Qwen) is used.
Unsloth 2025.2.5 patched 32 layers with 0 QKV layers, 0 O layers and 32 MLP layers.


### Data Prep
<a name="Data"></a>

We directly leverage [@willccbb](https://gist.github.com/willccbb/4676755236bb08cab5f4e54a0475d6fb) for data prep and all reward functions. You are free to create your own!

In [4]:
import re
from datasets import load_dataset, Dataset

# Load and prep dataset
SYSTEM_PROMPT = """
Given a git diff or patch you must analize and write a commit message.
The git commit message title should have max 60-character.
Let an empty line between title and description.
The git commit description should only have readable content, not code.

Respond in the following format:
<reasoning>
...
</reasoning>
<answer>
...
</answer>
"""

XML_COT_FORMAT = """\
<reasoning>
{reasoning}
</reasoning>
<answer>
{answer}
</answer>
"""

def extract_xml_answer(text: str) -> str:
    answer = text.split("<answer>")[-1]
    answer = answer.split("</answer>")[0]
    return answer.strip()

def extract_hash_answer(text: str) -> str | None:
    return text.replace("```git-commit-message", "").replace("```", "").strip()

# uncomment middle messages for 1-shot prompting
def get_gsm8k_questions(split = "train") -> Dataset:
    data = load_dataset('Tavernari/git-commit-message', 'default')[split] # type: ignore
    data = data.map(lambda x: { # type: ignore
        'prompt': [
            {'role': 'system', 'content': SYSTEM_PROMPT},
            {'role': 'user', 'content': x['input']}
        ],
        'answer': extract_hash_answer(x['output'])
    }) # type: ignore
    return data # type: ignore

dataset = get_gsm8k_questions()

# Reward functions
def correctness_reward_func(prompts, completions, answer, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    q = prompts[0][-1]['content']
    extracted_responses = [extract_xml_answer(r) for r in responses]
    print('-'*20, f"Question:\n{q}", f"\nAnswer:\n{answer[0]}", f"\nResponse:\n{responses[0]}", f"\nExtracted:\n{extracted_responses[0]}")
    return [2.0 if r == a else 0.0 for r, a in zip(extracted_responses, answer)]

def int_reward_func(completions, **kwargs) -> list[float]:
    responses = [completion[0]['content'] for completion in completions]
    extracted_responses = [extract_xml_answer(r) for r in responses]
    return [0.5 if r.isdigit() else 0.0 for r in extracted_responses]

def strict_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"^<reasoning>\n.*?\n</reasoning>\n<answer>\n.*?\n</answer>\n$"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def soft_format_reward_func(completions, **kwargs) -> list[float]:
    """Reward function that checks if the completion has a specific format."""
    pattern = r"<reasoning>.*?</reasoning>\s*<answer>.*?</answer>"
    responses = [completion[0]["content"] for completion in completions]
    matches = [re.match(pattern, r) for r in responses]
    return [0.5 if match else 0.0 for match in matches]

def count_xml(text) -> float:
    count = 0.0
    if text.count("<reasoning>\n") == 1:
        count += 0.125
    if text.count("\n</reasoning>\n") == 1:
        count += 0.125
    if text.count("\n<answer>\n") == 1:
        count += 0.125
        count -= len(text.split("\n</answer>\n")[-1])*0.001
    if text.count("\n</answer>") == 1:
        count += 0.125
        count -= (len(text.split("\n</answer>")[-1]) - 1)*0.001
    return count

def xmlcount_reward_func(completions, **kwargs) -> list[float]:
    contents = [completion[0]["content"] for completion in completions]
    return [count_xml(c) for c in contents]

<a name="Train"></a>
### Train the model

Now set up GRPO Trainer and all configurations!

In [5]:
from trl import GRPOConfig, GRPOTrainer
training_args = GRPOConfig(
    use_vllm = True, # use vLLM for fast inference!
    learning_rate = 5e-6,
    adam_beta1 = 0.9,
    adam_beta2 = 0.99,
    weight_decay = 0.1,
    warmup_ratio = 0.1,
    lr_scheduler_type = "cosine",
    optim = "paged_adamw_8bit",
    logging_steps = 1,
    bf16 = is_bfloat16_supported(),
    fp16 = not is_bfloat16_supported(),
    per_device_train_batch_size = 1,
    gradient_accumulation_steps = 1, # Increase to 4 for smoother training
    num_generations = 6, # Decrease if out of memory
    max_prompt_length = 2048,
    max_completion_length = 2048,
    num_train_epochs = 3, # Set to 1 for a full training run
    max_steps = 240,
    save_steps = 240,
    max_grad_norm = 0.1,
    report_to = "none", # Can use Weights & Biases
    output_dir = "outputs",
)

torch.distributed process group is initialized, but parallel_mode != ParallelMode.DISTRIBUTED. In order to use Torch DDP, launch your script with `python -m torch.distributed.launch


And let's run the trainer! If you scroll up, you'll see a table of rewards. The goal is to see the `reward` column increase!

You might have to wait 150 to 200 steps for any action. You'll probably get 0 reward for the first 100 steps. Please be patient!

| Step | Training Loss | reward    | reward_std | completion_length | kl       |
|------|---------------|-----------|------------|-------------------|----------|
| 1    | 0.000000      | 0.125000  | 0.000000   | 200.000000        | 0.000000 |
| 2    | 0.000000      | 0.072375  | 0.248112   | 200.000000        | 0.000000 |
| 3    | 0.000000      | -0.079000 | 0.163776   | 182.500000        | 0.000005 |


In [ ]:
trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        xmlcount_reward_func,
        soft_format_reward_func,
        strict_format_reward_func,
        int_reward_func,
        correctness_reward_func,
    ],
    args = training_args,
    train_dataset = dataset,
)
trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs = 1
   \\   /|    Num examples = 2,535 | Num Epochs = 1
O^O/ \_/ \    Batch size per device = 1 | Gradient Accumulation steps = 1
\        /    Total batch size = 1 | Total steps = 240
 "-____-"     Number of trainable parameters = 113,246,208


-------------------- Question:
diff --git a/src/main/kotlin/math/TwoSum.kt b/src/main/kotlin/math/TwoSum.kt
new file mode 100644
index 0000000..7b18f4d
--- /dev/null
+++ b/src/main/kotlin/math/TwoSum.kt
@@ -0,0 +1,19 @@
+package math
+/**
+ * Try all the pairs in the array and see if any of them add up to the target number.
+ * @param nums Array of integers.
+ * @param target Integer target.
+ * @return Indices of the two numbers such that they add up to target.
+ */
+fun twoSum(nums: IntArray, target: Int): IntArray{
+    for (index1 in nums.indices) {
+        val startIndex = index1 + 1
+        for (index2 in startIndex..nums.lastIndex) {
+            if (nums[index1] + nums[index2] == target) {
+                return intArrayOf(index1, index2)
+            }
+        }
+    }
+    return intArrayOf(0,1)
+
+}
\ No newline at end of file 
Answer:
Add TwoSum function for finding indices of two numbers that sum to target

This commit introduces the `twoSum` function, which takes an a

-------------------- Question:
diff --git a/src/merge.py b/src/merge.py
--- a/src/merge.py
+++ b/src/merge.py
@@ -13,5 +13,5 @@
         text = re.sub(r'[^\x00-\x7F]+', ' ', text)
         text = re.sub(r'\s+', ' ', text)
-        return text.strip().lower()
+        return text.strip()
     return text 
Answer:
Update text processing to preserve case in merge.py

This change modifies the text processing function to return the stripped text without converting it to lower case. Preserving the original case may be important for subsequent processing or user expectations, allowing for better handling of cases where case sensitivity is significant. 
Response:
<reasoning>
The provided patch shows a modification to the `merge.py` file. The original code included a line that lowercased the text and removed non-ASCII characters. The line was also stripped of leading and trailing whitespace. The updated code has removed the lowercasing, allowing for non-English text to pass through. 

The commi

<a name="Inference"></a>
### Inference
Now let's try the model we just trained! First, let's first try the model without any GRPO trained:

In [ ]:
text = tokenizer.apply_chat_template([
    {"role" : "user", "content" : """\
    diff --git forkSrcPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift forkDstPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift
new file mode 100644
index 0000000000000000000000000000000000000000..43875ea954cfda9cb9b28cded3c45473680c3382
--- /dev/null
+++ forkDstPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift
@@ -0,0 +1,45 @@
+@testable import Project
+import XCTest
+
+class ConfigurationTests: XCTestCase {
+    private let decoder = JSONDecoder()
+
+    // MARK: - Tests
+    func test_ConfigurationDecoding_WhenProfileCacheValuesWasProvided_UsesProvidedValues() throws {
+        // Given
+        let json = """
+        {
+            "version": "1.0.0",
+            "profileWithoutPinDuration": 3600,
+            "profileWithPinDuration": 86400
+        }
+        """
+
+        // When
+        let configuration = try decoder.decode(Configuration.self, from: Data(json.utf8))
+
+        // Then
+        XCTAssertEqual(configuration.version, "1.0.0")
+        XCTAssertEqual(configuration.profileWithoutPinDuration, 3600)
+        XCTAssertEqual(configuration.profileWithPinDuration, 86400)
+    }
+
+    func test_ConfigurationDecoding_WhenProfileCacheValuesNotProvided_UsesDefaultValue() throws {
+        // Given
+        let json = """
+        {
+            "version": "1.1.0"
+        }
+        """
+
+        // When
+        let configuration = try decoder.decode(Configuration.self, from: Data(json.utf8))
+
+        // Then
+        XCTAssertEqual(configuration.version, "1.1.0")
+        /// Default value is 30 minutes
+        XCTAssertEqual(configuration.profileWithoutPinDuration, 1800)
+        /// Default value is 12 hours
+        XCTAssertEqual(configuration.profileWithPinDuration, 43200)
+    }
+}
diff --git forkSrcPrefix/Project.xcodeproj/project.pbxproj forkDstPrefix/Project.xcodeproj/project.pbxproj
index 61b65d35abf98eda4c0845782bb0f1b58d5c89e5..f28dc9c7fae76b2076d59f25043974a39b32c68d 100644
--- forkSrcPrefix/Project.xcodeproj/project.pbxproj
+++ forkDstPrefix/Project.xcodeproj/project.pbxproj
@@ -2767,6 +2767,7 @@
 		F3D3D8D32AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8D22AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift */; };
 		F3D3D8F32AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8F22AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift */; };
 		F3D3D8F82AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8F72AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift */; };
+		F3D7C5C52D5B60BD00D8AA5D /* ConfigurationTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */; };
 		F3DD7E322BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3DD7E312BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift */; };
 		F3DFC0192C93387900FA7E2E /* EPGAssetModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3DFC0182C93387900FA7E2E /* EPGAssetModel.swift */; };
 		F3E04CDB2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3E04CDA2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift */; };
@@ -6495,6 +6496,7 @@
 		F3D3D8D22AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "BrowseViewController+DisplayInvalidatedCache.swift"; sourceTree = "<group>"; };
 		F3D3D8F22AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "InfoStorage+TokenValidityManager.swift"; sourceTree = "<group>"; };
 		F3D3D8F72AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = InfoStorageTokenExpirationManagerTests.swift; sourceTree = "<group>"; };
+		F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = ConfigurationTests.swift; sourceTree = "<group>"; };
 		F3DD7E312BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Dependency+DispatchQueueServiceName.swift"; sourceTree = "<group>"; };
 		F3DFC0182C93387900FA7E2E /* EPGAssetModel.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = EPGAssetModel.swift; sourceTree = "<group>"; };
 		F3E04CDA2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Browse.InteractiveSchedule+ViewController+DataDisplaying.swift"; sourceTree = "<group>"; };
@@ -15192,6 +15194,7 @@
 		C412C10929F162D5003C2D36 /* Configuration */ = {
 			isa = PBXGroup;
 			children = (
+				F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */,
 				C412C10A29F162DC003C2D36 /* ConfigurationVideoQualitySettingsTests.swift */,
 			);
 			path = Configuration;
@@ -22326,6 +22329,7 @@
 				A5E557262637264C00ED2159 /* Configuration.AppTransparency+Stub.swift in Sources */,
 				6B610F6F29B95C8500EF848A /* HeroControllableMock.swift in Sources */,
 				679F96CC2D4C0798000EA3AA /* ProductsPickerTrackerMock.swift in Sources */,
+				F3D7C5C52D5B60BD00D8AA5D /* ConfigurationTests.swift in Sources */,
 				9A82F2AE2B69161A006E35A1 /* SignInModels+Stub.swift in Sources */,
 				F3F99D192CB938FE0051AD9D /* TileOverrideImageSelectionStrategyHeroPresentationNoneTests.swift in Sources */,
 				671AA0372CEB858D00FA61BC /* DirectBillingMapperTests.swift in Sources */,
"""},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    [text],
    sampling_params = sampling_params,
    lora_request = None,
)[0].outputs[0].text

output

And now with the LoRA we just trained with GRPO - we first save the LoRA first!

In [ ]:
model.save_lora("grpo_saved_lora")

Now we load the LoRA and test:

In [ ]:
text = tokenizer.apply_chat_template([
    {"role" : "system", "content" : SYSTEM_PROMPT},
    {"role" : "user", "content" : """\
    diff --git forkSrcPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift forkDstPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift
new file mode 100644
index 0000000000000000000000000000000000000000..43875ea954cfda9cb9b28cded3c45473680c3382
--- /dev/null
+++ forkDstPrefix/ProjectTests/Tests/Configuration/ConfigurationTests.swift
@@ -0,0 +1,45 @@
+@testable import Project
+import XCTest
+
+class ConfigurationTests: XCTestCase {
+    private let decoder = JSONDecoder()
+
+    // MARK: - Tests
+    func test_ConfigurationDecoding_WhenProfileCacheValuesWasProvided_UsesProvidedValues() throws {
+        // Given
+        let json = """
+        {
+            "version": "1.0.0",
+            "profileWithoutPinDuration": 3600,
+            "profileWithPinDuration": 86400
+        }
+        """
+
+        // When
+        let configuration = try decoder.decode(Configuration.self, from: Data(json.utf8))
+
+        // Then
+        XCTAssertEqual(configuration.version, "1.0.0")
+        XCTAssertEqual(configuration.profileWithoutPinDuration, 3600)
+        XCTAssertEqual(configuration.profileWithPinDuration, 86400)
+    }
+
+    func test_ConfigurationDecoding_WhenProfileCacheValuesNotProvided_UsesDefaultValue() throws {
+        // Given
+        let json = """
+        {
+            "version": "1.1.0"
+        }
+        """
+
+        // When
+        let configuration = try decoder.decode(Configuration.self, from: Data(json.utf8))
+
+        // Then
+        XCTAssertEqual(configuration.version, "1.1.0")
+        /// Default value is 30 minutes
+        XCTAssertEqual(configuration.profileWithoutPinDuration, 1800)
+        /// Default value is 12 hours
+        XCTAssertEqual(configuration.profileWithPinDuration, 43200)
+    }
+}
diff --git forkSrcPrefix/Project.xcodeproj/project.pbxproj forkDstPrefix/Project.xcodeproj/project.pbxproj
index 61b65d35abf98eda4c0845782bb0f1b58d5c89e5..f28dc9c7fae76b2076d59f25043974a39b32c68d 100644
--- forkSrcPrefix/Project.xcodeproj/project.pbxproj
+++ forkDstPrefix/Project.xcodeproj/project.pbxproj
@@ -2767,6 +2767,7 @@
 		F3D3D8D32AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8D22AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift */; };
 		F3D3D8F32AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8F22AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift */; };
 		F3D3D8F82AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D3D8F72AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift */; };
+		F3D7C5C52D5B60BD00D8AA5D /* ConfigurationTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */; };
 		F3DD7E322BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3DD7E312BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift */; };
 		F3DFC0192C93387900FA7E2E /* EPGAssetModel.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3DFC0182C93387900FA7E2E /* EPGAssetModel.swift */; };
 		F3E04CDB2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift in Sources */ = {isa = PBXBuildFile; fileRef = F3E04CDA2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift */; };
@@ -6495,6 +6496,7 @@
 		F3D3D8D22AB8A5F000891143 /* BrowseViewController+DisplayInvalidatedCache.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "BrowseViewController+DisplayInvalidatedCache.swift"; sourceTree = "<group>"; };
 		F3D3D8F22AB9F1BE00891143 /* InfoStorage+TokenValidityManager.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "InfoStorage+TokenValidityManager.swift"; sourceTree = "<group>"; };
 		F3D3D8F72AB9F2C700891143 /* InfoStorageTokenExpirationManagerTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = InfoStorageTokenExpirationManagerTests.swift; sourceTree = "<group>"; };
+		F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = ConfigurationTests.swift; sourceTree = "<group>"; };
 		F3DD7E312BA86CF8003B7301 /* Dependency+DispatchQueueServiceName.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Dependency+DispatchQueueServiceName.swift"; sourceTree = "<group>"; };
 		F3DFC0182C93387900FA7E2E /* EPGAssetModel.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = EPGAssetModel.swift; sourceTree = "<group>"; };
 		F3E04CDA2C063C480054E143 /* Browse.InteractiveSchedule+ViewController+DataDisplaying.swift */ = {isa = PBXFileReference; lastKnownFileType = sourcecode.swift; path = "Browse.InteractiveSchedule+ViewController+DataDisplaying.swift"; sourceTree = "<group>"; };
@@ -15192,6 +15194,7 @@
 		C412C10929F162D5003C2D36 /* Configuration */ = {
 			isa = PBXGroup;
 			children = (
+				F3D7C5C42D5B60BD00D8AA5D /* ConfigurationTests.swift */,
 				C412C10A29F162DC003C2D36 /* ConfigurationVideoQualitySettingsTests.swift */,
 			);
 			path = Configuration;
@@ -22326,6 +22329,7 @@
 				A5E557262637264C00ED2159 /* Configuration.AppTransparency+Stub.swift in Sources */,
 				6B610F6F29B95C8500EF848A /* HeroControllableMock.swift in Sources */,
 				679F96CC2D4C0798000EA3AA /* ProductsPickerTrackerMock.swift in Sources */,
+				F3D7C5C52D5B60BD00D8AA5D /* ConfigurationTests.swift in Sources */,
 				9A82F2AE2B69161A006E35A1 /* SignInModels+Stub.swift in Sources */,
 				F3F99D192CB938FE0051AD9D /* TileOverrideImageSelectionStrategyHeroPresentationNoneTests.swift in Sources */,
 				671AA0372CEB858D00FA61BC /* DirectBillingMapperTests.swift in Sources */,
"""},
], tokenize = False, add_generation_prompt = True)

from vllm import SamplingParams
sampling_params = SamplingParams(
    temperature = 0.8,
    top_p = 0.95,
    max_tokens = 1024,
)
output = model.fast_generate(
    text,
    sampling_params = sampling_params,
    lora_request = model.load_lora("grpo_saved_lora"),
)[0].outputs[0].text

output

Our reasoning model is much better - it's not always correct, since we only trained it for an hour or so - it'll be better if we extend the sequence length and train for longer!

<a name="Save"></a>
### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16 or `merged_4bit` for int4. We also allow `lora` adapters as a fallback. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Merge to 16bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_16bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_16bit", token = "")

# Merge to 4bit
if False: model.save_pretrained_merged("model", tokenizer, save_method = "merged_4bit",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "merged_4bit", token = "")

# Just LoRA adapters
if False: model.save_pretrained_merged("model", tokenizer, save_method = "lora",)
if False: model.push_to_hub_merged("hf/model", tokenizer, save_method = "lora", token = "")

### GGUF / llama.cpp Conversion
To save to `GGUF` / `llama.cpp`, we support it natively now! We clone `llama.cpp` and we default save it to `q8_0`. We allow all methods like `q4_k_m`. Use `save_pretrained_gguf` for local saving and `push_to_hub_gguf` for uploading to HF.

Some supported quant methods (full list on our [Wiki page](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)):
* `q8_0` - Fast conversion. High resource use, but generally acceptable.
* `q4_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q4_K.
* `q5_k_m` - Recommended. Uses Q6_K for half of the attention.wv and feed_forward.w2 tensors, else Q5_K.

[**NEW**] To finetune and auto export to Ollama, try our [Ollama notebook](https://colab.research.google.com/drive/1WZDi7APtQ9VsvOrQSSC5DDtxq159j8iZ?usp=sharing)

In [ ]:
# Save to 8bit Q8_0
if False: model.save_pretrained_gguf("model", tokenizer,)
# Remember to go to https://huggingface.co/settings/tokens for a token!
# And change hf to your username!
if False: model.push_to_hub_gguf("hf/model", tokenizer, token = "")

# Save to 16bit GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "f16")
if False: model.push_to_hub_gguf("hf/model", tokenizer, quantization_method = "f16", token = "")

# Save to q4_k_m GGUF
if False: model.save_pretrained_gguf("model", tokenizer, quantization_method = "q4_k_m")
if True: model.push_to_hub_gguf("Tavernari/git-commit-message", tokenizer, quantization_method = "q4_k_m", token = "")

# Save to multiple GGUF options - much faster if you want multiple!
if False:
    model.push_to_hub_gguf(
        "hf/model", # Change hf to your username!
        tokenizer,
        quantization_method = ["q4_k_m", "q8_0", "q5_k_m",],
        token = "",
    )

Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Llama 3.2 Conversational notebook. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(1B_and_3B)-Conversational.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
